test/UnitWOPIUnlock.cpp

```diff
@@ -0,0 +1,167 @@
+/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 100 -*- */
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <config.h>
+
+#include "lokassert.hpp"
+#include <WopiTestServer.hpp>
+#include <Log.hpp>
+#include <helpers.hpp>
+#include <wsd/ClientSession.hpp>
+
+#include <Poco/Net/HTTPRequest.h>
+
+/// This is to test that we do not unlock before uploading
+/// the document, right before unloading the last view.
+class UnitWopiUnlock : public WopiTestServer
+{
+    STATE_ENUM(Phase, Load, Lock, Modify, Upload, Unlock, Done) _phase;
+
+    std::string _lockState;
+    std::string _lockString;
+    std::size_t _sessionCount;
+
+public:
+    UnitWopiUnlock()
+        : WopiTestServer("UnitWopiUnlock")
+        , _phase(Phase::Load)
+        , _lockState("UNLOCK")
+        , _sessionCount(0)
+    {
+    }
+
+    void configCheckFileInfo(Poco::JSON::Object::Ptr fileInfo) override
+    {
+        fileInfo->set("SupportsLocks", "true");
+    }
+
+    std::unique_ptr<http::Response>
+    assertPutFileRequest(const Poco::Net::HTTPRequest& /*request*/) override
+    {
+        LOG_TST("assertPutFileRequest");
+        LOK_ASSERT_STATE(_phase, Phase::Upload);
+
+        TRANSITION_STATE(_phase, Phase::Unlock);
+        return nullptr;
+    }
+
+    void assertLockRequest(const Poco::Net::HTTPRequest& request) override
+    {
+        const std::string lock = request.get("X-WOPI-Lock", std::string());
+        const std::string newLockState = request.get("X-WOPI-Override", std::string());
+        LOG_TST("In " << toString(_phase) << ", X-WOPI-Lock: " << lock << ", X-WOPI-Override: "
+                      << newLockState << ", for URI: " << request.getURI());
+
+        if (_phase == Phase::Lock)
+        {
+            LOK_ASSERT_EQUAL_MESSAGE("Expected X-WOPI-Override:LOCK", std::string("LOCK"),
+                                     newLockState);
+            LOK_ASSERT_MESSAGE("Lock String cannot be empty", !lock.empty());
+            _lockState = newLockState;
+            _lockString = lock;
+            TRANSITION_STATE(_phase, Phase::Modify);
+        }
+        else if (_phase == Phase::Unlock)
+        {
+            LOK_ASSERT_EQUAL_MESSAGE("Expected X-WOPI-Override:UNLOCK", std::string("UNLOCK"),
+                                     newLockState);
+            LOK_ASSERT_MESSAGE("Document is not unlocked", _lockState != "UNLOCK");
```

Comment:
shouldn't be   `_lockState == "LOCK"`   to be sure it's not empty?

Reply:
On line 61 we assert that `newLockState` is "LOCK" and then we assign it to `_lockState`. I suppose we could make it more explicit.